lib/crypto/client.cpp

```diff
@@ -378,6 +378,92 @@ OlmClient::create_outbound_session(const std::string &identity_key, const std::s
         return session;
 }
 
+SASPtr
+OlmClient::sas_init()
+{
+        auto sas        = create_olm_object<SASObject>();
+        auto random_buf = create_buffer(olm_create_sas_random_length(sas.get()));
+
+        const int ret = olm_create_sas(sas.get(), random_buf.data(), random_buf.size());
+
+        if (ret == -1)
+                throw olm_exception("create_sas_instance", sas.get());
+
+        return sas;
+}
+
+std::string
+OlmClient::sas_get_pub_key(OlmSAS *sas)
+{
+        auto pub_key_buffer = create_buffer(olm_sas_pubkey_length(sas));
+
+        const int ret = olm_sas_get_pubkey(sas, pub_key_buffer.data(), pub_key_buffer.size());
+
+        if (ret == -1)
+                throw olm_exception("get_public_key", sas);
+
+        return to_string(pub_key_buffer);
+}
+
+void
+OlmClient::set_their_key(OlmSAS *sas, std::string their_public_key)
+{
+        auto pub_key_buffer = to_binary_buf(their_public_key);
+
+        const int ret = olm_sas_set_their_key(sas, pub_key_buffer.data(), pub_key_buffer.size());
+
+        if (ret == -1)
+                throw olm_exception("get_public_key", sas);
+}
+
+void
+OlmClient::generate_bytes_decimal(OlmSAS *sas, std::string info, std::vector<int> &output_list)
```

Comment:
Why did you change this to an output parameter?

Reply:
Rather than outputting a vector I felt this was neater

Reply:
I'm not so sure about that. Having a function return a value, when it is used to get a value, is usually the idiomatic part. Especially since it make the usage code more complicated, i.e.:

```
        std::vector<int> alice_emoji;
        alice->generate_bytes_emoji(alice_sas.get(), info, alice_emoji);
```

vs

```
        std::vector<int> alice_emoji = alice->generate_bytes_emoji(alice_sas.get(), info);
```

Reply:
Yeah actually, I'll change it then